incubator/sparkoperator/templates/spark-operator.yaml

```diff
@@ -0,0 +1,170 @@
+        imagePullPolicy: Always
+        command: ["/usr/bin/gencerts.sh", "-n", "{{ .Values.operatorNamespace }}", "-p"]
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: sparkoperator-cleanup
+  annotations:
+    "helm.sh/hook": post-delete
+    "helm.sh/hook-delete-policy": hook-succeeded
+spec:
+  template:
+    spec:
+      restartPolicy: OnFailure
+      containers:
+      - name: main
+        image: {{ .Values.operatorImageName }}:{{ .Values.operatorVersion }}
+        imagePullPolicy: Always
+        command: ["/usr/bin/delete-resource.sh", "--kind", "namespaces", "--name", "{{ .Values.operatorNamespace }}"]
+{{ else }}
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: {{ .Values.operatorNamespace }}
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: sparkoperator
+  namespace: {{ .Values.operatorNamespace }}
+{{ end }}
+---
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: sparkoperator
+  namespace: {{ .Values.operatorNamespace }}
+  labels:
+    app.kubernetes.io/name: sparkoperator
+    app.kubernetes.io/version: {{ .Values.operatorVersion }}
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: sparkoperator
+      app.kubernetes.io/version: {{ .Values.operatorVersion }}
+  strategy:
+    type: Recreate
+  template:
+    metadata:
+      {{- if .Values.enableMetrics }}
+      annotations:
+        prometheus.io/scrape: "true"
+        prometheus.io/port: "10254"
+        prometheus.io/path: "/metrics"
+      {{- end }}
+      labels:
+        app.kubernetes.io/name: sparkoperator
+        app.kubernetes.io/version: {{ .Values.operatorVersion }}
+      initializers:
+        pending: []
+    spec:
+      serviceAccountName: sparkoperator
+      {{- if .Values.enableWebhook }}
+      volumes:
+        - name: webhook-certs
+          secret:
+            secretName: spark-webhook-certs
+      {{- end }}
+      containers:
+      - name: sparkoperator
+        image: {{ .Values.operatorImageName }}:{{ .Values.operatorVersion }}
+        imagePullPolicy: Always
+        {{- if .Values.enableWebhook }}
+        volumeMounts:
+          - name: webhook-certs
+            mountPath: /etc/webhook-certs
+        {{- end }}
+        {{- if .Values.enableMetrics }}
+        ports:
+          - containerPort: 10254
+        {{ end }}
+        command: ["/usr/bin/spark-operator"]
+        {{- if and (.Values.enableMetrics) (.Values.enableWebhook) }}
+        args:
+        - -enable-metrics=true
+        - -metrics-labels=app_type
+        - -enable-webhook=true
+        - -logtostderr
+        {{- else if (.Values.enableMetrics) }}
+        args:
+        - -enable-metrics=true
+        - -metrics-labels=app_type
+        - -logtostderr
+        {{- else if (.Values.enableWebhook) }}
+        args:
+        - -enable-webhook=true
+        - -logtostderr
+        {{- else }}
+        args:
+        - -logtostderr
+        {{- end }}
+---
+{{ if .Values.enableWebhook }}
```

Comment:
Strage, I get container-creating on spark operator, and error 
```
  Warning  FailedMount  1m (x9 over 3m)  kubelet, test-cluster-tnrgym2mzvyk-minion-0  MountVolume.SetUp failed for volume "webhook-certs" : secrets "spark-webhook-certs" not found
```

Reply:
Hmm, interesting looking at gen-certs job output
```
HTTP/1.1 403 Forbidden
Content-Type: application/json
X-Content-Type-Options: nosniff
Date: Wed, 22 Aug 2018 13:59:47 GMT
Content-Length: 331
Connection: close

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "secrets is forbidden: User \"system:serviceaccount:spark-operator:sparkoperator\" cannot create secrets in the namespace \"spark-operator\"",
  "reason": "Forbidden",
  "details": {
    "kind": "secrets"
  },
  "code": 403
}
```

Reply:
Weird. Working for me. I'll take a look today.

Reply:
I debug the issue, and service accounts are created, but roles and clusterroles not for some strange reason. But helm tells me that it is created.. maybe some issue with cluster

Reply:
Ok, they were created properly. Not sure what could be wrong.

Reply:
Interesting, running this batch job separately succeeded.

Reply:
Hmm it might be a Helm issue. Now I remember running `kubectl create clusterrolebinding bind-cluster-admin --clusterrole=cluster-admin --serviceaccount=kube-system:default
`. Can you try that too?

Reply:
@yuchaoran2011 yes, I will do one more experiment tomorrow on trying to force clusterrole and binding to be before a init job.